numpy/__init__.cython-30.pxd

```diff
@@ -1046,3 +1046,161 @@ cdef inline NPY_DATETIMEUNIT get_datetime64_unit(object obj) nogil:
+        # When buffering is enabled, grows the inner loop if possible
+        NPY_ITER_GROWINNER
+        # Delay allocation of buffers until first Reset* call
+        NPY_ITER_DELAY_BUFALLOC
+        # When NPY_KEEPORDER is specified, disable reversing negative-stride axes
+        NPY_ITER_DONT_NEGATE_STRIDES
+        NPY_ITER_COPY_IF_OVERLAP
+        # The operand will be read from and written to
+        NPY_ITER_READWRITE
+        # The operand will only be read from
+        NPY_ITER_READONLY
+        # The operand will only be written to
+        NPY_ITER_WRITEONLY
+        # The operand's data must be in native byte order
+        NPY_ITER_NBO
+        # The operand's data must be aligned
+        NPY_ITER_ALIGNED
+        # The operand's data must be contiguous (within the inner loop)
+        NPY_ITER_CONTIG
+        # The operand may be copied to satisfy requirements
+        NPY_ITER_COPY
+        # The operand may be copied with WRITEBACKIFCOPY to satisfy requirements
+        NPY_ITER_UPDATEIFCOPY
+        # Allocate the operand if it is NULL
+        NPY_ITER_ALLOCATE
+        # If an operand is allocated, don't use any subtype
+        NPY_ITER_NO_SUBTYPE
+        # This is a virtual array slot, operand is NULL but temporary data is there
+        NPY_ITER_VIRTUAL
+        # Require that the dimension match the iterator dimensions exactly
+        NPY_ITER_NO_BROADCAST
+        # A mask is being used on this array, affects buffer -> array copy
+        NPY_ITER_WRITEMASKED
+        # This array is the mask for all WRITEMASKED operands
+        NPY_ITER_ARRAYMASK
+        # Assume iterator order data access for COPY_IF_OVERLAP
+        NPY_ITER_OVERLAP_ASSUME_ELEMENTWISE
+
+    # construction and destruction functions
+    NpyIter* NpyIter_New(ndarray arr, npy_uint32 flags, NPY_ORDER order,
+                         NPY_CASTING casting, dtype datatype)
+    NpyIter* NpyIter_MultiNew(npy_intp nop, PyArrayObject** op, npy_uint32 flags,
+                              NPY_ORDER order, NPY_CASTING casting, npy_uint32*
+                              op_flags, PyArray_Descr** op_dtypes)
+    NpyIter* NpyIter_AdvancedNew(npy_intp nop, PyArrayObject** op,
+                                 npy_uint32 flags, NPY_ORDER order,
+                                 NPY_CASTING casting, npy_uint32* op_flags,
+                                 PyArray_Descr** op_dtypes, int oa_ndim,
+                                 int** op_axes, const npy_intp* itershape,
+                                 npy_intp buffersize)
+    NpyIter* NpyIter_Copy(NpyIter* it)
+    int NpyIter_RemoveAxis(NpyIter* it, int axis)
+    int NpyIter_RemoveMultiIndex(NpyIter* it)
+    int NpyIter_EnableExternalLoop(NpyIter* it)
+    int NpyIter_Deallocate(NpyIter* it)
+    int NpyIter_Reset(NpyIter* it, char** errmsg) nogil
+    int NpyIter_ResetToIterIndexRange(NpyIter* it, npy_intp istart,
+                                      npy_intp iend, char** errmsg) nogil
```

Comment:
Thanks, just a quick comment, since you are iterating.

I would be tempted to not add `nogil` as itis only correct when `errmsg` is passed, and I frankly find it a bit awkward.
(I am not sure it is actually important API for an unused threading usecase, or more of a try to allow use without GIL but without a concrete use-case)

My main question is how much you checked the correct `except ...` statements.  Pretty use `NpyIter_Dealloc` needs an `except 0`, so I think there may be a few more.  Also clearly creating a new one needs `except NULL`.  (in other words, we need to carefully review those)

Reply:
Just addressed the above comment in [`e5c2883` (#24691)](https://github.com/numpy/numpy/pull/24691/commits/e5c28830a23224594db7dd4791a31c85a675b39f) by removing the nogil and replacing it with `except? NPY_FAIL`. I also checked the implementation of of the functions listed and added appropriate `except ...` where needed. Please give another look when you can.